### PR TITLE
Demonstrate Parameters for HTTP::Client

### PR DESCRIPTION
**Pull request by HCLarsen**
Adding parameters to an HTTP request seems like a common use case, so some example code would be very useful in the documentation.

### CODE REVIEW

File: src/http/client.cr [hunk 1]
```diff
@@ -13,6 +13,20 @@
...
+# Parameters can be added to any request with the `HTTP::Params#encode` method, which
+# converts a `Hash` or `NamedTuple` to a URL encoded HTTP query.
+#
+# ```
+# require "http/client"
+#
+# params = HTTP::Params.encode({ "q" => "test"}) # => q=test
+# response = HTTP::Client.get "http://www.example.com?" + params
+# response.status_code      # => 200
+# response.body.lines.first # => "<!doctype html>"
```
Comment (oprypin) [inline src/http/client.cr]: No need to repeat this from the previous example if it's not relevant.

File: src/http/client.cr [hunk 2]
```diff
@@ -13,6 +13,20 @@
...
 #
+# ### Parameters
+#
+# Parameters can be added to any request with the `HTTP::Params#encode` method, which
+# converts a `Hash` or `NamedTuple` to a URL encoded HTTP query.
+#
+# ```
+# require "http/client"
+#
+# params = HTTP::Params.encode({ "q" => "test"}) # => q=test
```
Comment (oprypin) [inline src/http/client.cr]: No space in `{"`. Run `crystal tool format`, that might fix it.

File: src/http/client.cr [hunk 3]
```diff
@@ -13,6 +13,20 @@
...
+# response = HTTP::Client.get "http://www.example.com?" + params
```
Comment (oprypin) [inline src/http/client.cr]: Maybe worth having a more complex example? 2 parameters, one of which contains something that needs to be escaped? `{"author" => "John Doe", "offset" => 20}`

Comment (HCLarsen) [reply src/http/client.cr]: Ok, those sound good to me.

File: src/http/client.cr (hunk 2 above)
Comment (oprypin) [reply src/http/client.cr]: This is also what caused the CI failure.

File: src/http/client.cr (hunk 1 above)
Comment (HCLarsen) [reply src/http/client.cr]: I think it'd be a good idea to keep at least one of those output lines to show symmetry with the previous example.

File: src/http/client.cr (hunk 3 above)
Comment (HCLarsen) [reply src/http/client.cr]: Although I had to change the second value to a string instead of an int. "20" works, but 20 throws an error.

**Comment by RX14** [comment]
I strongly dislike using this method for creating queries with params. The correct way is to set the params using a `URI` object. We have to parse any `String` values into a `URI` anyway, so we should probably have a named param for the `String` version which sets the params on the URI.

**Comment by HCLarsen** [comment]
I like it. It's what I was advised to do when I asked about this use case in the gitter chat. I like that it keeps everything within the HTTP module. I also really like that you can take a hash, and just turn it into a URL encoded string in just one step. Very user friendly.

**Comment by oprypin** [comment]
Can you finally show the better replacement that you're talking about, in an example?

**Comment by HCLarsen** [comment]
I don't see any params method in the URI class, so I'm not sure what you're getting at. I agree that an example would help us understand what you're suggesting.

**Comment by RX14** [comment]
`HTTP::Client.get("http://www.example.com", query_params: {"foo" => "bar"})` would be ideal but currently you need to do
```cr
uri = URI.parse("http://www.example.com")
uri.query = HTTP::Params.encode({ "q" => "test"})
HTTP::Client.get(uri)
```

**Comment by HCLarsen** [comment]
I see how that works, but I don't think it's inherently better. Now granted, that example should be added to the documentation, but I'd place it in the URI class documentation. Quite frankly, having just looked through that documentation, I did not see anything to suggest this as possible.

**Comment by HCLarsen** [comment]
Also, I love the idea of `HTTP::Client.get("http://www.example.com", query_params: {"foo" => "bar"})`. Would you like me to add an issue for suggesting that?

**Comment by asterite** [comment]
What about:

```crystal
HTTP::Client.get("http://www.example.com?a=b", query_params: {"foo" => "bar"})
```

Or even:

```crystal
HTTP::Client.get("http://www.example.com?foo=baz", query_params: {"foo" => "bar"})
```

That means HTTP::Client now has to parse the URI, parse the query parameters and combine them with the given ones.

Please, let's not add more and more features for no reason. In fact, can we declare a feature freeze? :-)

**Comment by HCLarsen** [comment]
First, can someone tell me why the checks are failing? Not sure what I did wrong.

**Comment by HCLarsen** [comment]
@asterite, in that case, I would say that the developer is doing a very bad job by writing some params in the address, and others in the params hash. With proper documentation, any good developer would know to put all their params in the params hash, and NOT in the address string. In any case, it seems like a different discussion than this PR, so I'm game for opening an issue if you guys are.

**Comment by oprypin** [comment]
> First, can someone tell me why the checks are failing?

`crystal tool format` that I mentioned earlier

**Comment by HCLarsen** [comment]
Just did that, and it changed files that I haven't touched. Is it supposed to do that?

**Comment by HCLarsen** [comment]
When I ran `crystal tool format` it changed 

src/compiler/crystal/tools/init.cr
src/llvm/lib_llvm_ext.cr

As well as client.cr. Before I commit and push, can I be sure that it won't break anything?

**Comment by oprypin** [comment]
Dang... you might have to run the latest formatter on the latest source code.
`make`
`bin/crystal tool format`

**Comment by HCLarsen** [comment]
Make gave me an error.

`Could not locate llvm-config`

**Comment by oprypin** [comment]
Well, basically you need to do the proper setup for compiling Crystal, in order to contribute to it. But that's not necessary for now. The tool pointed out the problem with the file you changed. Just ignore the rest and move on.

**Comment by HCLarsen** [comment]
Like this?

**Comment by wontruefree** [comment]
@HCLarsen I think the is a good example.

**Comment by wontruefree** [comment]
This looks good to me.

**Comment by wontruefree** [comment]
@asterite I hope you dont mind me pinging you but I see you are the last one that posted a doc change.  Does this look good to you?

**Comment by wontruefree** [comment]
@HCLarsen what is the status on this?

**Comment by HCLarsen** [comment]
You're asking me? I fixed what I was told to fix, and yet there's been no movement on this in over a year.

**Comment by wontruefree** [comment]
@HCLarsen I dont know why it was not merged I would have liked more docs around this.

@rx14 can we merge this?

**Comment by HCLarsen** [comment]
I would like more docs around anything.

**Pull request merged**